presto-main/src/main/java/com/facebook/presto/operator/GroupIdOperator.java

```diff
@@ -0,0 +1,227 @@
+                        .appendNull()
+                        .build();
+            }
+
+            Block[] groupIdBlocks = new Block[groupingSetNullChannels.length];
+            for (int i = 0; i < groupingSetNullChannels.length; i++) {
+                BlockBuilder builder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), 1);
+                BIGINT.writeLong(builder, i);
+                groupIdBlocks[i] = builder.build();
+            }
+
+            return new GroupIdOperator(operatorContext, outputTypes, groupingSetNullChannels, nullBlocks, groupIdBlocks);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new GroupIdOperatorFactory(operatorId, planNodeId, inputTypes, groupingSetChannels);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> types;
+    private final BitSet[] groupingSetNullChannels;
+    private final Block[] nullBlocks;
+    private final Block[] groupIdBlocks;
+
+    private Page currentPage = null;
+    private Page outputPage = null;
+    private int currentGroupingSet = 0;
+    private boolean finishing;
+
+    public GroupIdOperator(
+            OperatorContext operatorContext,
+            List<Type> inputTypes,
+            BitSet[] groupingSetNullChannels,
+            Block[] nullBlocks,
+            Block[] groupIdBlocks)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = requireNonNull(inputTypes, "inputTypes is null");
+        this.groupingSetNullChannels =  requireNonNull(groupingSetNullChannels, "groupingSetNullChannels is null");
+        this.nullBlocks = requireNonNull(nullBlocks);
+        this.groupIdBlocks = requireNonNull(groupIdBlocks);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && currentPage == null;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && currentPage == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        checkState(currentPage == null, "currentPage must be null to add a new page");
+
+        currentPage = requireNonNull(page, "page is null");
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (outputPage == null && currentPage == null) {
+            return null;
+        }
+
+        if (outputPage == null) {
+            process();
+        }
+
+        Page oldOutputPage = outputPage;
```

Comment:
I don't think you need to store outputPage as a field? It looks like you always extract it fully after every process()